auth/ante_test.go

```diff
@@ -0,0 +1,367 @@
+	acc2 = happ.AccountKeeper.GetAccount(ctx, hmTypes.AccAddressToHeimdallAddress(addr2))
+	accNumber1 := acc1.GetAccountNumber()
+	accNumber2 := acc2.GetAccountNumber()
+
+	// test good tx from one signer
+	tx = types.NewTestTx(ctx, msg1, priv1, accNumber1, uint64(0))
+	checkValidTx(t, anteHandler, ctx, tx, false)
+
+	// test good tx from one signer
+	tx = types.NewTestTx(ctx, msg1, priv1, accNumber1, uint64(1))
+	checkValidTx(t, anteHandler, ctx, tx, false)
+
+	// // new tx from wrong account number
+	tx = types.NewTestTx(ctx, msg2, priv2, accNumber2, uint64(1))
+	checkInvalidTx(t, anteHandler, ctx, tx, false, sdk.CodeUnauthorized)
+
+	// from correct account number but wrong private key
+	tx = types.NewTestTx(ctx, msg2, priv1, accNumber2, uint64(0)) // with private key 1
+	checkInvalidTx(t, anteHandler, ctx, tx, false, sdk.CodeUnauthorized)
+
+	// from correct account number but wrong private key
+	tx = types.NewTestTx(ctx, msg2, priv2, accNumber2, uint64(0)) // with private key 2 (account 2)
+	checkValidTx(t, anteHandler, ctx, tx, false)
+}
+
+// Test logic around fee deduction.
+func (suite *AnteTestSuite) TestFees() {
+	t, happ, ctx, anteHandler := suite.T(), suite.app, suite.ctx, suite.anteHandler
+
+	// keys and addresses
+	priv1, _, addr1 := sdkAuth.KeyTestPubAddr()
+
+	// set the accounts
+	acc1 := happ.AccountKeeper.NewAccountWithAddress(ctx, hmTypes.AccAddressToHeimdallAddress(addr1))
+	happ.AccountKeeper.SetAccount(ctx, acc1)
+
+	// msg and signatures
+	var tx sdk.Tx
+	msg1 := sdkAuth.NewTestMsg(addr1)
+	acc1 = happ.AccountKeeper.GetAccount(ctx, hmTypes.AccAddressToHeimdallAddress(addr1))
+	tx = types.NewTestTx(ctx, msg1, priv1, uint64(0), uint64(0))
+	checkInvalidTx(t, anteHandler, ctx, tx, false, sdk.CodeInsufficientFunds)
+
+	// set some coins
+	acc1.SetCoins(sdk.NewCoins(sdk.NewInt64Coin(authTypes.FeeToken, 149)))
+	happ.AccountKeeper.SetAccount(ctx, acc1)
+	checkInvalidTx(t, anteHandler, ctx, tx, false, sdk.CodeInsufficientFunds)
+
+	require.True(t, happ.SupplyKeeper.GetModuleAccount(ctx, authTypes.FeeCollectorName).GetCoins().Empty())
+	require.True(sdk.IntEq(t, happ.AccountKeeper.GetAccount(ctx, hmTypes.AccAddressToHeimdallAddress(addr1)).GetCoins().AmountOf(authTypes.FeeToken), sdk.NewInt(149)))
+
+	amt, _ := sdk.NewIntFromString(authTypes.DefaultTxFees)
+	acc1.SetCoins(sdk.NewCoins(sdk.NewCoin(authTypes.FeeToken, amt)))
+	happ.AccountKeeper.SetAccount(ctx, acc1)
+	checkValidTx(t, anteHandler, ctx, tx, false)
+
+	require.True(sdk.IntEq(t, happ.SupplyKeeper.GetModuleAccount(ctx, types.FeeCollectorName).GetCoins().AmountOf(authTypes.FeeToken), amt))
+	require.True(sdk.IntEq(t, happ.AccountKeeper.GetAccount(ctx, hmTypes.AccAddressToHeimdallAddress(addr1)).GetCoins().AmountOf(authTypes.FeeToken), sdk.NewInt(0)))
+}
+
+//
+// utils
+//
+
+// run the tx through the anteHandler and ensure its valid
+func checkValidTx(t *testing.T, anteHandler sdk.AnteHandler, ctx sdk.Context, tx sdk.Tx, simulate bool) (sdk.Context, sdk.Result, bool) {
```

Comment:
too many inputs to func, should we use a struct instead? or make this a method on an existing struct?

Reply:
No need to. This is just an internal function. But yeah we can create a struct.

Reply:
Was hoping there is an existing transaction struct with the same functionality that could have been used. So sure makes sense to keep it the way it is.